worker/src/main/java/gov/cms/ab2d/worker/service/JobProcessingServiceImpl.java

```diff
@@ -30,28 +60,136 @@ public Job putJobInProgress(String jobId) {
+    public Job processJob(final String jobId) {
+
+        final Job job = jobRepository.findByJobUuid(jobId);
+        log.info("Found job : {} - {} - {} ", job.getId(), job.getJobUuid(), job.getStatus());
+
+        final Sponsor sponsor = job.getUser().getSponsor();
+
+        final List<Contract> attestedContracts = sponsor.getAttestedContracts();
+        log.info("number of attested contracts : {}", attestedContracts.size());
+
+
+        final Path outputDirPath = Paths.get(efsMount, job.getJobUuid());
+        final Path outputDir = fileService.createDirectory(outputDirPath);
+        for (Contract contract : attestedContracts) {
+            log.info("contract : {} ", contract.getContractNumber());
+
+            JobOutput jobOutput = null;
+            try {
+                jobOutput = processContract(outputDir, contract);
+                jobOutput.setError(false);
+            } catch (Exception e) {
+                jobOutput.setError(true);
+                log.error("error processing contract : {} ", contract.getContractNumber(), e);
+            }
+
+            job.addJobOutput(jobOutput);
+            jobOutputRepository.save(jobOutput);
+        }
+
+        completeJob(job);
+        return job;
+    }
+
+
+    private JobOutput processContract(final Path outputDir, Contract contract) {
+        final String filename = contract.getContractNumber() + NDJSON_EXTENSION;
+        final var ndJsonFile = fileService.createFile(outputDir, filename);
+
+        final var patientsByContract = beneficiaryAdapter.getPatientsByContract(contract.getContractNumber());
+
+        final var futureResourcesHandles =  new ArrayList<Future<List<Resource>>>();
+
+        int counter = 0;
+        for (var patient : patientsByContract.getPatients()) {
+            final var resources = bfdClientAdapter.getEobBundleResources(patient.getPatientId());
+            futureResourcesHandles.add(resources);
+
+            ++counter;
+            if (counter % GROUP_SIZE == 0) {
+                processResources(futureResourcesHandles, ndJsonFile);
+            }
+        }
+
+        processResources(futureResourcesHandles, ndJsonFile);
+
+        JobOutput jobOutput = new JobOutput();
+        jobOutput.setFilePath(getFilePath(ndJsonFile));
+        jobOutput.setFhirResourceType("ExplanationOfBenefits");
+
+        return jobOutput;
+    }
+
+    private String getFilePath(Path ndJsonFile) {
+        return ndJsonFile == null
+                ? null
+                : ndJsonFile.toAbsolutePath().toString();
+    }
+
+    private void processResources(List<Future<List<Resource>>> futureHandles, Path outputFile) {
+
+        final Iterator<Future<List<Resource>>> iterator = futureHandles.iterator();
+        while (iterator.hasNext()) {
+            final Future<List<Resource>> futureResources = iterator.next();
+
+            List<Resource> resources = null;
+            try {
+                resources = futureResources.get();
```

Comment:
Iterating over a list of `Futures` and calling `get` on each might not scale very well and hamper parallel processing, too. The problem here is that `get` is a _blocking_ method: it will be stuck until that particular Future is done. If you encounter one particularly slow Future, you will be stuck with it while others might have finished already.

At minimum, check for `isDone()` before calling `get`. But see my other comments: you might be better off refactoring the entire thing a bit.

Reply:
refined this section to fit better with the refactored logic in the comment below